articles/152_roslaunch_multi_machine_launch.md

```diff
@@ -0,0 +1,310 @@
+---
+layout: default
+title: ROS 2 Multi-Machine Launching
+permalink: articles/roslaunch_mml.html
+abstract:
+  Robotic systems are often distributed across multiple networked machines.
+  This document describes proposed modifications and enhancements to ROS2's
+  launch system to facilitate launching, monitoring, and shutting
+  down systems spread across multiple machines.
+author: '[Matt Lanting](https://github.com/mlanting)'
+published: false
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Authors: {{ page.author }}
+
+## Purpose
+
+Allow a system of ROS nodes to be launched on a hardware architecture that is
+spread across multiple networked computers and facilitate introspection and
+management of the system from a single machine.
+
+## Justification
+
+Nodes can need to run on different hosts for a variety of reasons.  Some possible
+use cases:
+
+- A large robot could with hosts located physically near the hardware they are
+controlling such as cameras or other sensors
+- A robot with hosts with different architectures in order to use specialized
+processing hardware
+- A robot with a cluster of machines that do distributed processing of data
+- A network of multiple virtual hosts for testing purposes
+- A swarm of independent drones that can cooperate but do not require
+communication with each other
+- A client computer is used to launch and monitor a system on a remote host but
+is not required for the system to operate
+
+## Capabilities
+
+In order to meet the above use cases, a launch system needs to have a number of
+capabilities, including:
+
+- Connecting to a remote host and running nodes on it
+- Pushing configuration parameters for nodes to remote hosts
+- Monitoring the status and tracking the lifecycle of nodes
+- Recovering from failures by optionally restarting nodes
+- Gracefully shutting down nodes on every host
+
+Most of these are just extensions of the design goals for the single-machine
+version of roslaunch for ROS2, but in addition to extending those goals to
+remote machines, we would also like to consider a few more advanced features,
+including:
+
+- Load balancing nodes on distributed networks
+- Command line tools for managing and monitoring systems across machines
+- Mechanisms for locating files and executables across machines
+- Sharing and synchronizing files across machines
```

Comment:
I see what you mean. Some of these such as `Command line tools for managing and monitoring systems across machines` is probably best suited by a community package than one provided by ros2.

Additionally, what is mean by "load balancing" in this case?

Reply:
in my opinion, i believe

- Mechanisms for locating files and executables across machines
- Sharing and synchronizing files across machines

these are totally off tipic from launch system, why this is integrated into launcher? could we think about these features more generic if necessary?

Reply:
I don't think I would say they're "totally off topic"; synchronizing a built workspace between machines is something I've had to deal with constantly when launch on multi-machine systems, and I've spent a considerably amount of time writing scripts to handle efficiently deploying large workspaces (I've got one that's 5.3 GB right now) across multiple ROS hosts.

But it is true that the concept of synchronizing files isn't tightly coupled to launching; there are people who will be interested in multi-machine launching but don't need to sync anything, and there are probably others who will want to sync data but aren't launching anything, so it might make more sense to break that out into its own system.

Reply:
Well, I think we need to consider what constitutes things for multi-machine launching vs. tools for multi-machine ros2 setups. Our best bet might be to consolidate the features that would help multi-machine launching in the `launch` command within this guide and then think about maybe a suite of packages containing features helpful to multi-machine setups.

Reply:
In ROS 1, files were looked up via something like "find file X in package Y", and in multi machine launching it was required that packages and files existed in both machines (maybe not in the same place, but they had to be discoverable).

That seemed to work ok, though obviously there are situations, especially in development, where this isn't ideal. However, I do think we'd do well to keep the tools "small and sharp" and deal with this problem outside of launch. If there appears to be a really good default way to solve this problem, then we can consider adding tool support.

However, I'd encourage you guys to get it working with the assumption that all the packages existing on the local and remote machines (or in the "remote" containers), and then we can workout mechanisms for synchronization or container setup as needed.